File: c++/0128-longest-consecutive-sequence/solution.cpp
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        if (nums.size() == 0) {
            return 0;
        }

        unordered_set<int> set(nums.begin(), nums.end());

        int longest = 1;
        for (auto& n : set) {
            if (!set.count(n - 1)) {
                int count = 1;

                while (set.count(n + count)) {
                    ++count;
                }

                longest = max(count, longest);
            }

            if (longest > nums.size() / 2) {
                break;
            }
        }

        return longest;
    }
};

File: java/0128-longest-consecutive-sequence/Solution.java
import java.util.HashSet;
import java.util.Set;

public class Solution {
    public int longestConsecutive(int[] nums) {
        if (nums.length == 0) {
            return 0;
        }

        Set<Integer> set = new HashSet<>();
        for (int n : nums) {
            set.add(n);
        }

        int longest = 1;
        for (int n : nums) {
            if (!set.contains(n - 1)) {
                int count = 1;

                while (set.contains(n + 1)) {
                    ++n;
                    ++count;
                }

                longest = Math.max(count, longest);
            }

            if (longest > nums.length / 2) {
                break;
            }
        }

        return longest;
    }
}
